File: android/pano/src/main/kotlin/com/malletcrafts/sitephotos/pano/Panorama.kt
package com.malletcrafts.sitephotos.pano

import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.sin
import kotlin.math.tan

/**
 * The 360 → cube-face projection, on the device.
 *
 * This is the second implementation of the formula in mallet_estimator's
 * panorama.py. It exists because both apps live on one phone and a site has no
 * signal: a face that has to travel to a server and back is a face ImageMeter
 * never sees while anyone is still standing in the room.
 *
 * Two implementations of one formula drift silently — nobody notices a face
 * two degrees off, they measure the wrong wall months later. So this file is
 * held to the goldens the Python side publishes
 * (mallet_estimator/tests/golden/), and both sides read the SAME source
 * panorama rather than each building its own.
 *
 * Deliberately free of android.graphics: it takes and returns plain int
 * arrays, so it runs — and is tested — on an ordinary JVM. The Android layer
 * converts Bitmap ↔ IntArray at the edges.
 */
object Panorama {

    const val DEFAULT_FOV = 110.0
    const val DEFAULT_FACE_PX = 1600

    const val FOV_MIN = 30.0
    const val FOV_MAX = 160.0
    const val FACE_PX_MIN = 256
    const val FACE_PX_MAX = 4096

    /** name, yaw°, pitch° — the same six, in the same order, as the server. */
    val FACES: List<Triple<String, Double, Double>> = listOf(
        Triple("front", 0.0, 0.0),
        Triple("right", 90.0, 0.0),
        Triple("back", 180.0, 0.0),
        Triple("left", 270.0, 0.0),
        Triple("up", 0.0, 90.0),
        Triple("down", 0.0, -90.0),
    )

    /** An equirectangular source: packed 0xRRGGBB, row-major, width*height. */
    class Image(val width: Int, val height: Int, val pixels: IntArray) {
        init {
            require(pixels.size == width * height) {
                "expected ${width * height} pixels, got ${pixels.size}"
            }
        }
    }

    /**
     * True when the image is plausibly a 2:1 equirectangular panorama.
     *
     * Splitting a normal photo produces convincing-looking garbage, which is
     * worse than refusing: nobody doubts a picture that looks like a room.
     */
    fun looksEquirect(width: Int, height: Int, tolerance: Double = 0.10): Boolean {
        if (width <= 0 || height <= 0) return false
        val ratio = width.toDouble() / height.toDouble()
        return kotlin.math.abs(ratio - 2.0) <= tolerance * 2.0
    }

    fun clampFov(fov: Double?): Double =
        (fov ?: DEFAULT_FOV).coerceIn(FOV_MIN, FOV_MAX)

    fun clampFacePx(px: Int?): Int =
        (px ?: DEFAULT_FACE_PX).coerceIn(FACE_PX_MIN, FACE_PX_MAX)

    /**
     * Sample one gnomonic face. Bilinear; longitude WRAPS (the back face
     * straddles the ±180° seam) and latitude CLAMPS at the poles.
     */
    fun faceFromEquirect(
        pano: Image,
        yawDeg: Double,
        pitchDeg: Double,
        fovDeg: Double,
        facePx: Int,
    ): Image {
        val w = pano.width
        val h = pano.height
        val lam0 = Math.toRadians(yawDeg)
        val phi0 = Math.toRadians(pitchDeg)

        // Camera basis. `up` is the analytic ∂forward/∂pitch, which stays
        // valid at the poles where crossing with world-up degenerates — and
        // the poles are exactly where the ceiling and floor faces point.
        val fx = cos(phi0) * sin(lam0)
        val fy = sin(phi0)
        val fz = cos(phi0) * cos(lam0)
        val rx = cos(lam0)
        val ry = 0.0
        val rz = -sin(lam0)
        val ux = -sin(phi0) * sin(lam0)
        val uy = cos(phi0)
        val uz = -sin(phi0) * cos(lam0)

        val t = tan(Math.toRadians(fovDeg) / 2.0)
        // linspace(-t, t, facePx): endpoints INCLUSIVE, which is why the step
        // divides by facePx - 1. Getting this wrong shifts every face by half
        // a pixel — invisible by eye, and a permanent disagreement with the
        // server's copy.
        val step = if (facePx > 1) (2.0 * t) / (facePx - 1) else 0.0

        val out = IntArray(facePx * facePx)
        for (row in 0 until facePx) {
            val b = t - row * step                 // top → bottom
            for (col in 0 until facePx) {
                val a = -t + col * step            // left → right

                val dx = fx + a * rx + b * ux
                val dy = fy + a * ry + b * uy
                val dz = fz + a * rz + b * uz

                val lam = atan2(dx, dz)
                val phi = atan2(dy, hypot(dx, dz))

                // Angle → source pixel, sampling at pixel centres.
                val x = (lam / (2.0 * Math.PI) + 0.5) * w - 0.5
                val y = (0.5 - phi / Math.PI) * h - 0.5

                val x0 = floor(x).toInt()
                val y0 = floor(y).toInt()
                val wx = (x - x0).toFloat()
                val wy = (y - y0).toFloat()

                val x0m = Math.floorMod(x0, w)
                val x1m = Math.floorMod(x0 + 1, w)
                val y0c = y0.coerceIn(0, h - 1)
                val y1c = (y0 + 1).coerceIn(0, h - 1)

                val p00 = pano.pixels[y0c * w + x0m]
                val p10 = pano.pixels[y0c * w + x1m]
                val p01 = pano.pixels[y1c * w + x0m]
                val p11 = pano.pixels[y1c * w + x1m]

                val c00 = (1 - wx) * (1 - wy)
                val c10 = wx * (1 - wy)
                val c01 = (1 - wx) * wy
                val c11 = wx * wy

                var packed = 0
                for (shift in intArrayOf(16, 8, 0)) {
                    val v = ((p00 ushr shift and 0xFF) * c00 +
                             (p10 ushr shift and 0xFF) * c10 +
                             (p01 ushr shift and 0xFF) * c01 +
                             (p11 ushr shift and 0xFF) * c11)
                    // +0.5 then truncate, matching the server's rounding.
                    val bch = (v + 0.5f).toInt().coerceIn(0, 255)
                    packed = packed or (bch shl shift)
                }
                out[row * facePx + col] = packed
            }
        }
        return Image(facePx, facePx, out)
    }

    /** All six faces, keyed by name. */
    fun splitEquirect(
        pano: Image,
        fov: Double? = null,
        facePx: Int? = null,
    ): Map<String, Image> {
        require(looksEquirect(pano.width, pano.height)) {
            "not a 2:1 equirectangular panorama: ${pano.width}x${pano.height}"
        }
        val f = clampFov(fov)
        val px = clampFacePx(facePx)
        return FACES.associate { (name, yaw, pitch) ->
            name to faceFromEquirect(pano, yaw, pitch, f, px)
        }
    }
}

File: mallet_estimator/panorama.py
# Equirectangular 360 → flat faces (gnomonic projection). Pure layer: numpy +
# Pillow only, no frappe — unit-testable without a bench (TESTING.md layer 1).
#
# A site 360 is the SOURCE OF TRUTH; the six faces are derived artifacts,
# regenerable at any FOV. The default 110° (not the cube-map 90°) is a
# deliberate annotation choice: faces sit 90° apart, so 110° gives ~20° of
# overlap on every edge — a defect near a wall corner shows on BOTH adjacent
# faces with context. The price is that the faces no longer reassemble into a
# seamless cube; an interactive viewer renders from the equirect instead.
import io
import math

import numpy as np
from PIL import Image

DEFAULT_FOV = 110.0
DEFAULT_FACE_PX = 1600

# Face name → (yaw°, pitch°). Yaw 0 = wherever the camera's front was; the
# names are camera-relative, not compass-relative, and that is fine because
# annotation happens per photo, not across photos.
FACES = (
    ("front", 0.0, 0.0),
    ("right", 90.0, 0.0),
    ("back", 180.0, 0.0),
    ("left", 270.0, 0.0),
    ("up", 0.0, 90.0),
    ("down", 0.0, -90.0),
)
FACE_NAMES = tuple(f[0] for f in FACES)

# A flat photograph — a phone snap of one wall, taken instead of a 360 — is
# its own single "face". It is NOT one of the six and must never be projected
# like one, but everything downstream of the split (annotating it, attaching a
# marked-up copy, filing it against a face) treats it exactly the same way.
# The Android side writes this same token (FaceWriter.PHOTO_FACE).
PHOTO_FACE = "photo"

# What a face token is allowed to be wherever an ANNOTATION is concerned.
# Kept apart from FACE_NAMES on purpose: the splitter must still see six.
ANNOTATABLE_FACES = FACE_NAMES + (PHOTO_FACE,)

# Guard rails: below 30° a face is a keyhole, above 160° the gnomonic plane
# stretches to uselessness (180° is mathematically infinite).
FOV_MIN, FOV_MAX = 30.0, 160.0
FACE_PX_MIN, FACE_PX_MAX = 256, 4096

# Panos wider than this are downscaled before sampling so one 100-megapixel
# upload can't blow a shared bench worker's memory. 12k × 6k stays plenty for
# 1600 px faces.
MAX_PANO_WIDTH = 12000


def clamp_params(fov=None, face_px=None):
    fov = float(fov or DEFAULT_FOV)
    face_px = int(face_px or DEFAULT_FACE_PX)
    return (min(max(fov, FOV_MIN), FOV_MAX),
            min(max(face_px, FACE_PX_MIN), FACE_PX_MAX))


def looks_equirect(width, height, tolerance=0.10):
    """A full equirectangular pano is exactly 2:1. Anything far off is a normal
    photo or a cropped pano — splitting it would produce convincing-looking
    garbage, which is worse than refusing."""
    if not width or not height:
        return False
    return abs(width / height - 2.0) <= 2.0 * tolerance


def face_from_equirect(pano, yaw_deg, pitch_deg, fov_deg, face_px):
    """Sample one gnomonic face out of an equirect uint8 array (H×W×3).
    Bilinear, longitude wraps (the 'back' face straddles the ±180° seam),
    latitude clamps at the poles."""
    h, w = pano.shape[:2]
    lam0 = math.radians(yaw_deg)
    phi0 = math.radians(pitch_deg)

    # Camera basis. `up` is the analytic ∂forward/∂pitch, which stays valid at
    # the poles where cross-with-world-up would degenerate.
    forward = np.array([math.cos(phi0) * math.sin(lam0),
                        math.sin(phi0),
                        math.cos(phi0) * math.cos(lam0)], dtype=np.float64)
    right = np.array([math.cos(lam0), 0.0, -math.sin(lam0)], dtype=np.float64)
    up = np.array([-math.sin(phi0) * math.sin(lam0),
                   math.cos(phi0),
                   -math.sin(phi0) * math.cos(lam0)], dtype=np.float64)

    t = math.tan(math.radians(fov_deg) / 2.0)
    a = np.linspace(-t, t, face_px, dtype=np.float64)      # left → right
    b = np.linspace(t, -t, face_px, dtype=np.float64)      # top → bottom
    A, B = np.meshgrid(a, b)
    d = (forward[None, None, :]
         + A[..., None] * right[None, None, :]
         + B[..., None] * up[None, None, :])

    lam = np.arctan2(d[..., 0], d[..., 2])
    phi = np.arctan2(d[..., 1], np.hypot(d[..., 0], d[..., 2]))

    # Angle → source pixel, sampling at pixel centres.
    x = (lam / (2.0 * np.pi) + 0.5) * w - 0.5
    y = (0.5 - phi / np.pi) * h - 0.5

    x0 = np.floor(x).astype(np.int64)
    y0 = np.floor(y).astype(np.int64)
    wx = (x - x0).astype(np.float32)[..., None]
    wy = (y - y0).astype(np.float32)[..., None]
    x0m, x1m = np.mod(x0, w), np.mod(x0 + 1, w)
    y0c = np.clip(y0, 0, h - 1)
    y1c = np.clip(y0 + 1, 0, h - 1)

    # Gather the four neighbours as uint8 and only widen the gathered pixels —
    # never the whole pano — to keep peak memory a few face-sized buffers.
    out = (pano[y0c, x0m].astype(np.float32) * (1 - wx) * (1 - wy)
           + pano[y0c, x1m].astype(np.float32) * wx * (1 - wy)
           + pano[y1c, x0m].astype(np.float32) * (1 - wx) * wy
           + pano[y1c, x1m].astype(np.float32) * wx * wy)
    return np.clip(out + 0.5, 0, 255).astype(np.uint8)


def split_equirect(image_bytes, fov=None, face_px=None):
    """The feature: 360 bytes in, {face_name: PIL.Image} out."""
    fov, face_px = clamp_params(fov, face_px)
    img = Image.open(io.BytesIO(image_bytes)).convert("RGB")
    if not looks_equirect(img.width, img.height):
        raise ValueError(
            f"not an equirectangular 360 photo: {img.width}×{img.height} "
            f"(expected width ≈ 2 × height)")
    if img.width > MAX_PANO_WIDTH:
        img = img.resize((MAX_PANO_WIDTH, MAX_PANO_WIDTH // 2), Image.LANCZOS)
    pano = np.asarray(img, dtype=np.uint8)
    return {
        name: Image.fromarray(face_from_equirect(pano, yaw, pitch, fov, face_px))
        for name, yaw, pitch in FACES
    }


def split_to_jpeg(image_bytes, fov=None, face_px=None, quality=90):
    """{face_name: JPEG bytes} — what the frappe layer attaches as Files.
    JPEG, not PNG: these are photographs, and six 1600² PNGs of a building
    site would cost megabytes each for nothing."""
    out = {}
    for name, img in split_equirect(image_bytes, fov=fov, face_px=face_px).items():
        buf = io.BytesIO()
        img.save(buf, format="JPEG", quality=quality, optimize=True)
        out[name] = buf.getvalue()
    return out
